Enable support for FIDO2/U2F security keys

martelletto opened the pull request:
This PR enables client-side support for FIDO2/U2F security keys, allowing the generation and use of `ecdsa-sk` and `ed25519-sk` SSH keys on cross-platform and platform authenticators, from PowerShell and WSL.

Build and test instructions can be found at [here](https://gist.github.com/martelletto/6a7cf806c6433ac9ce71d66afb94d70e). I have uploaded  [Win64]( https://drive.google.com/file/d/1c3QnoF6WMFZfTStvk4banxyMwvjfxhA2/) and [ARM64](https://drive.google.com/file/d/1u9SCQHW7Pt1xt6HDXHfbeWT1KSDHLK3f/) symbols. (GitHub wouldn't let me attach them to the PR.)

Thank you,

-p.

CC @akshayku

martelletto committed:
sk-usbhid: use fido_{assert,cred}_set_clientdata()

martelletto committed:
build ssh-sk-helper

martelletto committed:
build definitions for libfido2

martelletto committed:
gitignore bits

martelletto committed:
enable ssh-sk

martelletto committed:
ssh-sk-client: use posix_spawn() to run ssh-sk-helper

martelletto committed:
ssh-keygen: don't prompt for FIDO PINs by default on Windows

only prompt for a FIDO PIN if sshsk_enroll() fails with
SSH_ERR_KEY_WRONG_PASSPHRASE, which never happens when sk-usbhid
uses webauthn.dll and means we are communicating directly with
a security key.

martelletto committed:
ssh: don't prompt for FIDO PINs by default on Windows

only prompt for a FIDO PIN if sshsk_sign() fails with
SSH_ERR_KEY_WRONG_PASSPHRASE, which never happens when sk-usbhid
uses webauthn.dll and means we are communicating directly with
a security key.

martelletto committed:
README blurb

martelletto committed:
sk-usbhid: always prefer windows://hello if device == NULL

martelletto committed:
sk-usbhid: randomise user_id by default

requested by akshayku@

martelletto committed:
fetch libfido2 using an auxiliary script

ghost commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/PowerShell/openssh-portable?pullRequest=541) <br/>All CLA requirements met.

martelletto committed:
sshsig: don't prompt for FIDO PINs by default on Windows

only prompt for a FIDO PIN if sshsig_wrap_sign() fails with
SSH_ERR_KEY_WRONG_PASSPHRASE, which never happens when sk-usbhid
uses webauthn.dll and means we are communicating directly with
a security key.

martelletto committed:
ssh-keygen -s: don't prompt for FIDO PINs by default on Windows

only prompt for a FIDO PIN if sshkey_certify() fails with
SSH_ERR_KEY_WRONG_PASSPHRASE, which never happens when sk-usbhid
uses webauthn.dll and means we are communicating directly with
a security key.

martelletto commented:
I have pushed two commits to add support for git commit signatures using FIDO2/U2F security keys. In addition to the two commits, you will need a version of git built with https://github.com/git-for-windows/git/pull/3561.

bagajjal commented on contrib/win32/openssh/GetFIDO2.ps1:
```diff
@@ -0,0 +1,63 @@
+
+$zip_path = Join-Path $PSScriptRoot "libfido2.zip"
+
+$release_url = "https://ambientworks.net/tmp/libfido2-1.10-b32020cc-win.zip"
```
I know this is a temporary change till we have build pipeline on our side to compile the libfido code.
Do you have any plans to release a newer version? If yes, can you refer to the release URL instead of https://ambientworks.net/tmp

martelletto replied:
IIRC, the plan was to vendor libfido2 (like we do with LibreSSL and zlib) due to legal reasons. The next release of libfido2 is scheduled for late January / early February.

bagajjal commented on contrib/win32/openssh/config.h.vs:
```diff
@@ -137,6 +137,17 @@
+#define ENABLE_SK_INTERNAL
+
+/* Define for discoverable credential support */
+#define HAVE_FIDO_CRED_PROT
```
Looks like these are not used anywhere?

martelletto replied:
`HAVE_FIDO_CRED_PROT` is used in [sk-usbhid.c:47](https://github.com/martelletto/openssh-portable/blob/fido2/sk-usbhid.c#L47). `ENABLE_SK_INTERNAL` is used in [readconf.c:2484](https://github.com/martelletto/openssh-portable/blob/fido2/readconf.c#L2484), [sk-usbhid.c:21](https://github.com/martelletto/openssh-portable/blob/fido2/sk-usbhid.c#L21), [ssh-add.c:799](https://github.com/martelletto/openssh-portable/blob/fido2/ssh-add.c#L799), [ssh-keygen.c:3384](https://github.com/martelletto/openssh-portable/blob/fido2/ssh-keygen.c#L3384), and [ssh-sk.c:117](https://github.com/martelletto/openssh-portable/blob/fido2/ssh-sk.c#L117).

bagajjal commented on contrib/win32/openssh/paths.targets:
```diff
@@ -6,11 +6,17 @@
     <LibreSSL-arm64-Path>$(SolutionDir)\LibreSSL\bin\desktop\arm64\</LibreSSL-arm64-Path>
     <LibreSSL-arm-Path>$(SolutionDir)\LibreSSL\bin\desktop\arm\</LibreSSL-arm-Path>
+    <fido2-Path>$(SolutionDir)\libfido2\</fido2-Path>
+    <fido2-x86-Path>$(SolutionDir)\libfido2\Win32\Release\v143\static\</fido2-x86-Path>
```
v143 -> I'm guessing it's compiled with visual studio 2022.
Win32-openssh is compiled with visual studio 2015, windows SDK v8.1. 

Does it have any compatibility issues?

bagajjal replied:
Can FIDO be compiled with visual studio 2015, windows SDK v10? 
We might switch to windows SDK v10 from windows SDK v8.1.

martelletto replied:
I am not aware of any incompatibility. I had to disable a warning (`C4201; nonstandard extension used: nameless struct/union`) to build libfido2 with Windows SDK 8.1 (using Visual Studio 2022). I don't have Visual Studio 2015 on my laptop.

martelletto replied:
Building with Windows SDK 8.1 using Visual Studio 2015 works as well.

bagajjal replied:
Can you please remove the v143 here. Having this makes us change this as we move to different version of visual studio

martelletto replied:
We can change it once the vendoring process is complete, as it needs to match the SDK version used to build libfido2. CC @anmenaga

bagajjal commented on sk-usbhid.c:
```diff
@@ -686,7 +686,7 @@ sk_enroll(uint32_t alg, const uint8_t *challenge, size_t challenge_len,
 	}
 	*enroll_response = NULL;
-	memset(user_id, 0, sizeof(user_id));
+	arc4random_buf(user_id, sizeof(user_id));
```
Any reason why this is modified?
If this is specific to windows then please add #ifdef WINDOWS block

martelletto replied:
This modification was requested by @akshayku. I have pushed 1e3d6d4a to put it under `#ifdef WINDOWS`.

akshayku replied:
This is needed so that existing credentials are not overwritten in FIDO Authenticators. I believe this should be generic to all the platforms but I am fine if we want to put it under #ifdef WINDOWS block.

bagajjal replied:
Please add a comment for future reference.

martelletto replied:
Addressed in 78920431.

bagajjal commented on sk-usbhid.c:
```diff
@@ -706,8 +706,10 @@ sk_enroll(uint32_t alg, const uint8_t *challenge, size_t challenge_len,
-	else
-		sk = sk_probe(NULL, NULL, 0);
+	else {
+		if ((sk = sk_open("windows://hello")) == NULL)
```
Please add #ifdef WINDOWS block for all changes specific to windows

bagajjal replied:
Please fix this for all changes in this file as this file is common between Windows and Linux.

bagajjal replied:
Also changes in ssh-keygen.c

martelletto replied:
Addressed in 7606c43ff507dcc3b483c6803573dfe5ca2da567 and b6c982fa26371506edb0ed4549d621ef757b1ec0.

bagajjal commented on ssh-sk-client.c:
```diff
@@ -43,15 +43,117 @@
+	char *ep;
+
+	memset(module_path, 0, sizeof(module_path));
+	if ((n = GetModuleFileNameA(NULL, module_path,
```
Please change it to GetModuleFileNameW()

martelletto replied:
Implemented in 6b54b8dee954f45899f2ae3defc86577e4d2231f. Please note that the resulting wide-character string still needs to be converted to a multi-byte string, since that is what `posix_spawnp()` understands, so we don't get much out of `GetModuleFileNameW()` besides additional complexity.

bagajjal commented on ssh-sk-client.c:
```diff
@@ -43,15 +43,117 @@
+	char *ep;
+	DWORD n;
+
+	if ((helper = getenv("SSH_SK_HELPER")) == NULL || strlen(helper) == 0) {
```
Where does the SSH_SK_HELPER environment variable set?

Also line 81, _PATH_SSH_SK_HELPER is referring to the unix path so I suggest to check if the helper exists or not before exiting this method.

martelletto replied:
The `SSH_SK_HELPER` environment variable is meant to be defined by the user as an optional override mechanism. Regarding `_PATH_SSH_SK_HELPER`, I have pushed ed39b666b95b12e8952b91016fd589218163c95e defining it as `C:\Windows\System32\OpenSSH\ssh-sk-helper.exe` on Windows.

bagajjal commented on ssh-sk-client.c:
```diff
@@ -43,15 +43,117 @@
+}
+
 static int
 start_helper(int *fdp, pid_t *pidp, void (**osigchldp)(int))
```
Instead of duplicating this function between windows and Linux, can you have one function and have windows specific changes in #ifdef WINDOWS block. The advantage of this approach is any changes made to this function from upstream repository (unix repository) gets reflected on the windows side as well

martelletto replied:
Addressed in dee6c5e7. In addition to my earlier comment on potential upstream preferences, the resulting sea of `#ifdef` blocks is a great place for subtle bugs to lurk, whereas a single block of code tends to be easier to reason about.

bagajjal commented on sshconnect2.c:
```diff
@@ -1281,8 +1281,13 @@ identity_sign(struct identity *id, u_char **sigp, size_t *lenp,
 		sign_key = prv;
 		if (sshkey_is_sk(sign_key)) {
+#ifdef WINDOWS
+			if ((sign_key->sk_flags &
```
This check seems to be opposite of Unix logic. Could you please add comments

martelletto replied:
This is indeed confusing. I have pushed 783697f2 with a clarification comment.

bagajjal commented on sshsig.c:
```diff
@@ -555,6 +558,55 @@ hash_file(int fd, const char *hashalg, struct sshbuf **bp)
 
+#ifdef WINDOWS
+int
+sshsig_sign_fd(struct sshkey *key, const char *hashalg,
```
same comment as earlier, can you please have the same function (instead of duplicating) and have the windows specific logic under #ifdef WINDOWS block.

martelletto replied:
Implemented in 3dd1a019. Please note that this probably reduces the likelihood of these changes being merged upstream, as the preference there is for fewer ifdefs where possible. I don't speak for the OpenSSH project, though -- this is pure speculation on my part.

bagajjal reviewed (changes requested):
Appreciate your time for adding this functionality.

martelletto committed:
sk-usbhid: +#ifdef WINDOWS; requested by bagajjal@

martelletto committed:
sshsig: merge sshsig_sign_fd(), requested by bagajjal@

martelletto committed:
ssh-sk-client: merge start_helper(), requested by bagajjal@

martelletto committed:
identity_sign: add a clarification comment

martelletto committed:
pathnames: adjust _PATH_SSH_SK_HELPER on Windows

martelletto committed:
ssh-sk-client: use GetModuleFileNameW(), requested by bagajjal@

martelletto committed:
fix previous

martelletto committed:
ssh-keygen: move Windows-specific bits under #ifdef WINDOWS; requested by bagajjal@

martelletto committed:
sk-usbhid: move Windows-specific bits under #ifdef WINDOWS; requested by bagajjal@

tavrez commented:
Have you implemented `sk_load_resident_keys`?

martelletto commented:
@tavrez `sk_load_resident_keys` requires FIDO 2.1 Credential Management, which isn't abstracted by webauthn.h AFAIK. It should work when talking directly to the authenticator (in administrator mode or older versions of Windows), but won't work otherwise.

tavrez commented:
webauthn.dll inside Windows 11 is version 3, I think it supports FIDO 2.1(according to [this commit](https://github.com/microsoft/webauthn/commit/8488eed5dc3ab4f166efec8d416aa6c5319e251a)).
But I haven't found anything related to getting stored keys inside it.

martelletto committed:
config.h: define HAVE_FIDO_DEV_GET_TOUCH_{BEGIN,STATUS}

these are needed by sk-usbhid.c when operating in environments
with multiple authenticators attached.

martelletto committed:
ssh-add: don't consider a sk_provider a key constraint

unless explicitly specified, don't consider a sk_provider a
key constraint, allowing ssh-sk keys using the default internal
provider to be added with SSH2_AGENTC_ADD_IDENTITY instead of
SSH2_AGENTC_ADD_ID_CONSTRAINED.

martelletto committed:
ssh-agent: promote sk_provider == NULL to "internal"

a sk_provider is required by ssh-sk-helper. as such, treat ssh-sk
keys without a provider as belonging to the "internal" provider.

martelletto commented:
I have included two commits (7393b48 and b31d77e) adding ssh-agent support for ecdsa-sk/ed25519-sk.

martelletto committed:
README: mention FIDO2 ssh-agent support

bagajjal commented on contrib/win32/openssh/OpenSSHBuildHelper.psm1:
```diff
@@ -344,7 +344,7 @@ function Start-OpenSSHPackage
 
     $buildDir = Join-Path $repositoryRoot ("bin\" + $folderName + "\" + $Configuration)
     $payload =  "sshd.exe", "ssh.exe", "ssh-agent.exe", "ssh-add.exe", "sftp.exe"
-    $payload += "sftp-server.exe", "scp.exe", "ssh-shellhost.exe", "ssh-keygen.exe", "ssh-keyscan.exe" 
```
ssh-sk-helper doesn't have the build dependencies defined. This leads to the build errors.
Please move the ssh-sk-helper into the "core" folder.

2>LINK : fatal error LNK1181: cannot open input file 'libssh.lib'
2>Done building project "ssh-sk-helper.vcxproj" -- FAILED.

![image](https://user-images.githubusercontent.com/23668037/145895089-4c18199c-e890-4813-b57a-b1c70b7ff0ad.png)

martelletto replied:
Addressed in ebc802a3. I hadn't tried (at all) to build from Visual Studio, and am a bit surprised that's the only change required. Thanks for pointing it out.

bagajjal replied:
I still see ssh-sk-helper ouside of "core" where rest of the project files are defined. Please make this change.

bagajjal replied:
can you move the ssh-sk-helper project under "core" similar to other projects like ssh, sshd, etc.

<img width="947" alt="image" src="https://user-images.githubusercontent.com/23668037/146289127-4ad39b1c-2cff-46aa-adfb-5b02b882e2c3.png">

martelletto replied:
Sure; addressed in f6e5ad4c.

bagajjal commented on contrib/win32/win32compat/ssh-agent/keyagent-request.c:
```diff
@@ -230,8 +231,14 @@ static int sign_blob(const struct sshkey *pubkey, u_char ** sig, size_t *siglen,
 
-	if (sshkey_private_deserialize(tmpbuf, &prikey) != 0 ||
-	    sshkey_sign(prikey, sig, siglen, blob, blen, algo, NULL, NULL, 0) != 0) {
+	if (sshkey_private_deserialize(tmpbuf, &prikey) != 0) {
```
Please have a windows block and move line 211 inside the windows block.

bagajjal commented on contrib/win32/win32compat/ssh-agent/keyagent-request.c:
```diff
@@ -230,8 +231,14 @@ static int sign_blob(const struct sshkey *pubkey, u_char ** sig, size_t *siglen,
-	if (sshkey_private_deserialize(tmpbuf, &prikey) != 0 ||
-	    sshkey_sign(prikey, sig, siglen, blob, blen, algo, NULL, NULL, 0) != 0) {
+	if (sshkey_private_deserialize(tmpbuf, &prikey) != 0) {
+		debug("cannot deserialize key");
```
why not log it as error?

bagajjal commented on contrib/win32/win32compat/ssh-agent/keyagent-request.c:
```diff
@@ -230,8 +231,14 @@ static int sign_blob(const struct sshkey *pubkey, u_char ** sig, size_t *siglen,
+		sk_provider = "internal";
+	if (sshkey_sign(prikey, sig, siglen, blob, blen, algo, sk_provider,
+	    NULL, 0) != 0) {
 		debug("cannot sign using retrieved key");
```
can you make it as error

bagajjal commented on sk-usbhid.c:
```diff
@@ -721,9 +732,9 @@ sk_enroll(uint32_t alg, const uint8_t *challenge, size_t challenge_len,
 		goto out;
 	}
-	if ((r = fido_cred_set_clientdata_hash(cred, challenge,
+	if ((r = fido_cred_set_clientdata(cred, challenge,
```
please have #ifdef WINDOWS block

martelletto replied:
Addressed in 6bd15af42.

bagajjal commented on sk-usbhid.c:
```diff
@@ -985,17 +995,25 @@ sk_sign(uint32_t alg, const uint8_t *data, size_t datalen,
 	*sign_response = NULL;
 	if (check_sign_load_resident_options(options, &device) != 0)
 		goto out; /* error already logged */
-	/* hash data to be signed before it goes to the security key */
```
Please move this to #ifndef WINDOWS block and also add a comment why this is not applicable for windows.

martelletto replied:
Addressed in 6bd15af424.

bagajjal commented on sk-usbhid.c:
```diff
@@ -1004,9 +1022,9 @@ sk_sign(uint32_t alg, const uint8_t *data, size_t datalen,
 		skdebug(__func__, "fido_assert_new failed");
 		goto out;
 	}
-	if ((r = fido_assert_set_clientdata_hash(assert, message,
```
please add #ifdef WINDOWS block, add a reason for the deviation.

martelletto replied:
Addressed in 6bd15af424.

bagajjal commented on .gitignore:
```diff
@@ -46,6 +46,7 @@ tags
 /config.h
 /contrib/win32/openssh/LibreSSL
 /contrib/win32/openssh/ZLib
+/contrib/win32/openssh/libfido2
```
I'm getting an error while generating the ecdsa-sk. It didn't prompt me to touch the yubikey. Although it didn't ask, I keep touching the yubikey and running the same command but it still fails.

![image](https://user-images.githubusercontent.com/23668037/145897200-e0627cad-78b0-4a4e-a242-d63413e88f5b.png)

bagajjal replied:
fyi, I'm compiling using the latest commit on my windows 10 machine.

![image](https://user-images.githubusercontent.com/23668037/145897686-e943585a-ad43-4f41-b672-4c1ab00e699e.png)

martelletto replied:
Looks like a build problem. Are you able to run ssh-sk-helper.exe as a standalone process and converse with it?

```
> .\ssh-sk-helper.exe -vvv
debug3: w32_fcntl ERROR: bad fd: 1
debug3: ssh_msg_recv entering
foo
ssh_msg_recv: read: bad msg_len 1718578954
ssh_msg_recv failed
```

If that doesn't work, please try rebuilding with Win32-OpenSSH.sln from ebc802a36, which works locally for me. If running ssh-sk-helper.exe works but ssh-keygen.exe still fails, can you please retry with debugging enabled?

```
> $Env:FIDO_DEBUG = 1
> .\ssh-keygen -vvv -t ecdsa-sk
```

Thank you,

-p.

martelletto committed:
sk-usbhid: clarify arc4random_buf() usage in sk_enroll()

martelletto committed:
sk-usbhid: more comments, ifdef blocks

no functional change

martelletto committed:
Win32-OpenSSH.sln: specify ssh-sk-helper's dependencies

bagajjal commented on sk-usbhid.c:
```diff
@@ -721,12 +733,31 @@ sk_enroll(uint32_t alg, const uint8_t *challenge, size_t challenge_len,
 	}
-	if ((r = fido_cred_set_clientdata_hash(cred, challenge,
-	    challenge_len)) != FIDO_OK) {
+#ifndef WINDOWS
```
Is there a reason why you are adding line 737 to 741 as it's new code for Unix?

https://github.com/openssh/openssh-portable/blob/V_8_6/sk-usbhid.c#L724

martelletto replied:
I used V_8_8 as reference: https://github.com/openssh/openssh-portable/blob/V_8_8/sk-usbhid.c#L724. I could trim that chunk to reduce the difference against V_8_6, but we would have to revisit it in V_8_8 or later. Let me know if you prefer that approach.

bagajjal commented on sk-usbhid.c:
```diff
@@ -1050,7 +1113,6 @@ sk_sign(uint32_t alg, const uint8_t *data, size_t datalen,
 	response = NULL;
 	ret = 0;
  out:
-	explicit_bzero(message, sizeof(message));
```
any reason why this is removed?

martelletto replied:
It's dead code on Windows. I've re-added it under #ifndef WINDOWS in 843d4d79.

martelletto committed:
Win32-OpenSSH.sln: move ssh-sk-helper to 'core'

martelletto committed:
sk-usbhid: another #ifndef WINDOWS chunk

martelletto committed:
empty commit to trigger ci

martelletto committed:
(another) empty commit to trigger ci

martelletto commented:
CI is failing with:
> Build execution time has reached the maximum allowed time for your plan (60 minutes).

Successful runs seem to take ~54 minutes, so we are very close to the limit. I could keep pushing empty commits until we are lucky and CI passes again, but perhaps it would make sense to bump the limit.

martelletto committed:
another attempt to get ci to pass

anmenaga commented:
> CI is failing with:
> 
> > Build execution time has reached the maximum allowed time for your plan (60 minutes).
> 
> Successful runs seem to take ~54 minutes, so we are very close to the limit. I could keep pushing empty commits until we are lucky and CI passes again, but perhaps it would make sense to bump the limit.

I've sent them a request to increase it to 90 minutes. Hopefully it gets approved.
